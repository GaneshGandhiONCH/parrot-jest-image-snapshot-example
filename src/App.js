/*
 * Copyright (c) 2018 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

import React, { Component } from 'react';

const OPEN_WEATHER_MAP_API_KEY = 'PLACE_KEY_HERE'; // obvs wouldnt keep this on the client outside of an example like this one
const PHOENIX_ZIP_CODE = '85013';

const GRADIENTS = {
  COOL: 'linear-gradient(to top, #0c3483 0%, #a2b6df 100%, #6b8cce 100%, #a2b6df 100%)',
  HOT: 'linear-gradient(to top, #f83600 0%, #f9d423 100%)',
  NEUTRAL: 'linear-gradient(to top, #29323c 0%, #485563 100%)',
};

class App extends Component {
  state = {
    loading: true,
    error: false,
    phxTemperature: '',
  };

  async componentDidMount() {
    try {
      const phxWeatherResponse = await fetch(
        `/weather?zip=${PHOENIX_ZIP_CODE},us&units=imperial&appid=${OPEN_WEATHER_MAP_API_KEY}`
      );
      const phxWeather = await phxWeatherResponse.json();
      const phxTemperature = phxWeather.main.temp;
      this.setState({ phxTemperature, loading: false });
    } catch (e) {
      this.setState({ loading: false, error: true });
    }
  }

  render() {
    const { phxTemperature, loading, error } = this.state;

    let background = GRADIENTS.NEUTRAL;
    let wittyMessageAboutTemperature;
    if (loading) {
      wittyMessageAboutTemperature = 'Let me find out...';
    } else if (error) {
      wittyMessageAboutTemperature = 'I am not sure...';
    } else if (phxTemperature > 90) {
      background = GRADIENTS.HOT;
      wittyMessageAboutTemperature = `Yes, it is ${phxTemperature} degrees but it is a dry heat.`;
    } else {
      background = GRADIENTS.COOL;
      wittyMessageAboutTemperature = `Nope! It is ${phxTemperature} degrees. Perfect snowbird weather.`;
    }

    return (
      <div
        style={{
          position: 'absolute',
          display: 'grid',
          height: '100%',
          width: '100%',
          color: 'white',
          background,
        }}
      >
        <div style={{ margin: 'auto', textAlign: 'center' }}>
          <h3>Is it hot in Phoenix right now?</h3>
          <p>{wittyMessageAboutTemperature}</p>
        </div>
      </div>
    );
  }
}

export default App;
