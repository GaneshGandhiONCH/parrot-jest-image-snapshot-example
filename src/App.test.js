/*
 * Copyright (c) 2018 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

import React from 'react';
import Adapter from 'enzyme-adapter-react-16';

import { createSerializer } from 'enzyme-to-json';

import App from './App';

import { configure } from 'enzyme';
import { mount } from 'enzyme';

configure({ adapter: new Adapter() });
expect.addSnapshotSerializer(createSerializer());

global.fetch = require('jest-fetch-mock');

beforeEach(() => {
  fetch.resetMocks();
});

it('renders a friendly message while loading', async () => {
  const fetchMock = fetch.mockResponse(JSON.stringify({ main: { temp: 101 } }));
  const wrapper = mount(<App />);
  // intentionally not waiting on componentDidMount to resolve as we want 'loading' state

  expect(wrapper).toMatchSnapshot();
});

it('renders a friendly message when there is an error fetching weather API', async () => {
  const fetchMock = fetch.mockReject('Error');
  const wrapper = mount(<App />);
  await expect(wrapper.instance().componentDidMount()).resolves.toBe(undefined);
  wrapper.update();

  expect(wrapper).toMatchSnapshot();
});

it('renders a message about how hot it is if temperature is above 90 degrees', async () => {
  const fetchMock = fetch.mockResponse(JSON.stringify({ main: { temp: 101 } }));
  const wrapper = mount(<App />);
  await expect(wrapper.instance().componentDidMount()).resolves.toBe(undefined);
  wrapper.update();

  expect(wrapper).toMatchSnapshot();
});

it('renders a message about how nice it is if temperature is below 90 degrees', async () => {
  const fetchMock = fetch.mockResponse(JSON.stringify({ main: { temp: 73 } }));
  const wrapper = mount(<App />);
  await expect(wrapper.instance().componentDidMount()).resolves.toBe(undefined);
  wrapper.update();

  expect(wrapper).toMatchSnapshot();
});
