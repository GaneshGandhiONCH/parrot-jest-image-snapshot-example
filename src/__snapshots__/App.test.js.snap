// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a friendly message when there is an error fetching weather API 1`] = `
<App>
  <div
    style={
      Object {
        "background": "linear-gradient(to top, #29323c 0%, #485563 100%)",
        "color": "white",
        "display": "grid",
        "height": "100%",
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "margin": "auto",
          "textAlign": "center",
        }
      }
    >
      <h3>
        Is it hot in Phoenix right now?
      </h3>
      <p>
        I am not sure...
      </p>
    </div>
  </div>
</App>
`;

exports[`renders a friendly message while loading 1`] = `
<App>
  <div
    style={
      Object {
        "background": "linear-gradient(to top, #29323c 0%, #485563 100%)",
        "color": "white",
        "display": "grid",
        "height": "100%",
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "margin": "auto",
          "textAlign": "center",
        }
      }
    >
      <h3>
        Is it hot in Phoenix right now?
      </h3>
      <p>
        Let me find out...
      </p>
    </div>
  </div>
</App>
`;

exports[`renders a message about how hot it is if temperature is above 90 degrees 1`] = `
<App>
  <div
    style={
      Object {
        "background": "linear-gradient(to top, #f83600 0%, #f9d423 100%)",
        "color": "white",
        "display": "grid",
        "height": "100%",
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "margin": "auto",
          "textAlign": "center",
        }
      }
    >
      <h3>
        Is it hot in Phoenix right now?
      </h3>
      <p>
        Yes, it is 101 degrees but it is a dry heat.
      </p>
    </div>
  </div>
</App>
`;

exports[`renders a message about how nice it is if temperature is below 90 degrees 1`] = `
<App>
  <div
    style={
      Object {
        "background": "linear-gradient(to top, #0c3483 0%, #a2b6df 100%, #6b8cce 100%, #a2b6df 100%)",
        "color": "white",
        "display": "grid",
        "height": "100%",
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "margin": "auto",
          "textAlign": "center",
        }
      }
    >
      <h3>
        Is it hot in Phoenix right now?
      </h3>
      <p>
        Nope! It is 73 degrees. Perfect snowbird weather.
      </p>
    </div>
  </div>
</App>
`;
